src/linear_algebra/matrix/fpow.lean

```diff
@@ -0,0 +1,397 @@
+calc A ^ (n - 1) = A ^ (n - 1) * A * A⁻¹ : by rw [mul_assoc, mul_eq_mul A, mul_nonsing_inv _ h,
+                                                  mul_one]
+             ... = A^n * A⁻¹             : by rw [← fpow_add_one h, sub_add_cancel]
+
+lemma fpow_add {A : M} (ha : is_unit A.det) (m n : ℤ) : A ^ (m + n) = A ^ m * A ^ n :=
+begin
+  induction n using int.induction_on with n ihn n ihn,
+  case hz : { simp },
+  { simp only [← add_assoc, fpow_add_one ha, ihn, mul_assoc] },
+  { rw [fpow_sub_one ha, ← mul_assoc, ← ihn, ← fpow_sub_one ha, add_sub_assoc] }
+end
+
+lemma fpow_add' {A : M} {m n : ℤ} (h : is_unit A.det ∨ (m < 0 ∧ n < 0 ∨ 0 ≤ m ∧ 0 ≤ n)) :
+  A ^ (m + n) = A ^ m * A ^ n :=
+begin
+  by_cases ha : is_unit A.det,
+  { exact fpow_add ha m n },
+  { casesI is_empty_or_nonempty n' with H H,
+    { ext i j,
+      exact is_empty_elim i },
+    nontriviality K,
+    simp only [false_or, ha] at h,
+    rcases h with (⟨hm', hn'⟩|⟨hm', hn'⟩),
+    { obtain ⟨k, rfl⟩ := exists_eq_neg_of_nat hm'.le,
+      obtain ⟨l, rfl⟩ := exists_eq_neg_of_nat hn'.le,
+      simp only [right.neg_neg_iff, coe_nat_pos] at hm' hn',
+      rw [←neg_add, ←int.coe_nat_add, fpow_neg_coe_nat, fpow_neg_coe_nat,
+          nonsing_inv_apply_not_is_unit, nonsing_inv_apply_not_is_unit, zero_mul],
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        simpa using hm' },
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        rw ←zero_add 0,
+        exact add_lt_add hm' hn' } },
+    { obtain ⟨k, rfl⟩ := eq_coe_of_zero_le hm',
+      obtain ⟨l, rfl⟩ := eq_coe_of_zero_le hn',
+      rw [←int.coe_nat_add, fpow_coe_nat, fpow_coe_nat, fpow_coe_nat, pow_add] } }
+end
+
+lemma fpow_add'' {A : M} {m n : ℤ} (h : is_unit A.det ∨ (m ≤ 0 ∧ n ≤ 0 ∨ 0 < m ∧ 0 < n)) :
+  A ^ (m + n) = A ^ m * A ^ n :=
+begin
+  by_cases ha : is_unit A.det,
+  { exact fpow_add ha m n },
+  { casesI is_empty_or_nonempty n' with H H,
+    { ext i j,
+      exact is_empty_elim i },
+    nontriviality K,
+    simp only [false_or, ha] at h,
+    rcases h with (⟨hm', hn'⟩|⟨hm', hn'⟩),
+    { rcases hm'.eq_or_lt with rfl|hk,
+      { rw [zero_add, gpow_zero, one_mul] },
+      rcases hn'.eq_or_lt with rfl|hl,
+      { rw [add_zero, gpow_zero, mul_one] },
+      obtain ⟨k, rfl⟩ := exists_eq_neg_of_nat hk.le,
+      obtain ⟨l, rfl⟩ := exists_eq_neg_of_nat hl.le,
+      simp only [right.neg_neg_iff, coe_nat_pos] at hk hl,
+      rw [←neg_add, ←int.coe_nat_add, fpow_neg_coe_nat, fpow_neg_coe_nat,
+          nonsing_inv_apply_not_is_unit, nonsing_inv_apply_not_is_unit, zero_mul],
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        simpa using hk },
+      { rwa [det_pow, is_unit_pos_pow_iff],
+        rw ←zero_add 0,
+        exact add_lt_add hk hl } },
+    { obtain ⟨k, rfl⟩ := eq_coe_of_zero_le hm'.le,
+      obtain ⟨l, rfl⟩ := eq_coe_of_zero_le hn'.le,
+      rw [←int.coe_nat_add, fpow_coe_nat, fpow_coe_nat, fpow_coe_nat, pow_add] } }
+end
```

Comment:
Not sure why `m, n` need to be strictly negative, you need to know whether `n'` is nonempty or `K` is nontrivial, the proof goes through without all those assumptions:
```suggestion
lemma fpow_add' {A : M} {m n : ℤ} (h : is_unit A.det ∨ (m ≤ 0 ∧ n ≤ 0 ∨ 0 ≤ m ∧ 0 ≤ n)) :
  A ^ (m + n) = A ^ m * A ^ n :=
begin
  by_cases ha : is_unit A.det,
  { exact fpow_add ha m n },
  { simp only [ha, false_or] at h,
    rcases h with ⟨hm', hn'⟩ | ⟨hm', hn'⟩,
    { obtain ⟨k, rfl⟩ := exists_eq_neg_of_nat hm',
      obtain ⟨l, rfl⟩ := exists_eq_neg_of_nat hn',
      cases k,
      { simp },
      rw [←neg_add, ←int.coe_nat_add, fpow_neg_coe_nat, fpow_neg_coe_nat,
          nonsing_inv_apply_not_is_unit, nonsing_inv_apply_not_is_unit, zero_mul],
      { rwa [det_pow, is_unit_pos_pow_iff k.succ_pos] },
      { rwa [det_pow, nat.succ_add, is_unit_pos_pow_iff (k + l).succ_pos] } },
    { obtain ⟨k, rfl⟩ := eq_coe_of_zero_le hm',
      obtain ⟨l, rfl⟩ := eq_coe_of_zero_le hn',
      rw [←int.coe_nat_add, fpow_coe_nat, fpow_coe_nat, fpow_coe_nat, pow_add] } }
end
```

Reply:
Indeed, why not let the person invoking the lemmas do the case distinction?
```suggestion
lemma fpow_add_of_nonpos {A : M} {m n : ℤ} (hm : m ≤ 0) (hn : n ≤ 0) :
  A ^ (m + n) = A ^ m * A ^ n :=
begin
  by_cases ha : is_unit A.det,
  { exact fpow_add ha m n },
  { obtain ⟨k, rfl⟩ := exists_eq_neg_of_nat hm,
    obtain ⟨l, rfl⟩ := exists_eq_neg_of_nat hn,
    cases k,
    { simp },
    rw [←neg_add, ←int.coe_nat_add, fpow_neg_coe_nat, fpow_neg_coe_nat,
        nonsing_inv_apply_not_is_unit, nonsing_inv_apply_not_is_unit, zero_mul],
    { rwa [det_pow, is_unit_pos_pow_iff k.succ_pos] },
    { rwa [det_pow, nat.succ_add, is_unit_pos_pow_iff (k + l).succ_pos] } },
end

lemma fpow_add_of_nonneg {A : M} {m n : ℤ} (hm : 0 ≤ m) (hn : 0 ≤ n) :
A ^ (m + n) = A ^ m * A ^ n :=
begin
  by_cases ha : is_unit A.det,
  { exact fpow_add ha m n },
  { obtain ⟨k, rfl⟩ := eq_coe_of_zero_le hm,
    obtain ⟨l, rfl⟩ := eq_coe_of_zero_le hn,
    rw [←int.coe_nat_add, fpow_coe_nat, fpow_coe_nat, fpow_coe_nat, pow_add] }
end
```